tensorflow_addons/layers/multihead_attention.py

```diff
@@ -0,0 +1,128 @@
+        output_size = self.output_size if self.output_size else num_value_features
+
+        self.query_kernel = self.add_weight(
+            name="query_kernel",
+            shape=[self.num_heads, num_query_features, self.head_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+        self.key_kernel = self.add_weight(
+            name="key_kernel",
+            shape=[self.num_heads, num_key_features, self.head_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+        self.value_kernel = self.add_weight(
+            name="value_kernel",
+            shape=[self.num_heads, num_value_features, self.head_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+        self.projection_kernel = self.add_weight(
+            name="projection_kernel",
+            shape=[self.num_heads, self.size_per_head, output_size],
+            # initializer=self.kernel_initializer,
+            # regularizer=self.kernel_regularizer,
+            # constraint=self.kernel_constraint,
+        )
+
+        if self.use_projection_bias:
+            self.projection_bias = self.add_weight(
+                name="projection_bias",
+                shape=[output_size],
+                # initializer=self.bias_initializer,
+                # regularizer=self.bias_regularizer,
+                # constraint=self.bias_constraint,
+            )
+        else:
+            self.projection_bias = None
+
+        self.dropout = tf.keras.layers.Dropout(self.dropout_rate)
+
+        super().build(input_shape)
+
+    def call(self, inputs):
+
+        # Nomenclature for einsum
+        # ------------------------
+        # N = query elements
+        # M = key/value elements
+        # H = heads
+        # I = input features
+        # O = output features
+
+        query = inputs[0]
+        key = inputs[1]
+        value = inputs[2]
+        attention_mask = inputs[3] if len(inputs) == 4 else None
+
+        # Linear transformations
+        query = tf.einsum("...NI , HIO -> ...NHO", query, self.query_kernel)
+        key = tf.einsum("...MI , HIO -> ...MHO", key, self.key_kernel)
+        value = tf.einsum("...MI , HIO -> ...MHO", value, self.value_kernel)
+
+        # Scale dot-product, doing the division to either query or key
+        # instead of their product saves some computation
+        depth = tf.cast(tf.shape(query)[-1], tf.float32)
+        query /= tf.sqrt(depth)
+
+        # Calculate dot product attention
+        logits = tf.einsum("...NHO,...MHO->...HNM", key, query)
+
+        # apply mask
+        if attention_mask is not None:
+            logits += -10e9 * (1.0 - attention_mask)
```

Comment:
Maybe `attention_mask` should be casted to float32.

Reply:
Nice catch :)